### ISSUE TITLE
Automatic check and warnings of possible stability issues

### ISSUE THREAD

**Issue by sigvaldm** [puncproject/PUNCpp#4]
Several stability criteria exists, perhaps the most common ones:

dx*wps << 1 for all species s
dx on the order or smaller than lambda_Ds for all species s

but it's also often said that velocities should mostly be less than dx/dt (although I have some doubts to which extent this is really necessary for a electrostatic code). This means:

dx*wpe < dx / lambda_De

**Comment by sigvaldm** [comment]
For reference, I formulate some stability criteria. They are all based on resolving physical characterstic lengths (L) and times (T) by characteristic lengths (dx) and times (dt) of the simulation:

dx/L ~ 1
dt/T ~ 1

dt can only be the time-step of course, but one can consider several characteristic lengths dx for the mesh, for instance N-root of max/min cell volume or voronoi volume (as used in PTetra). Perhaps equally correct are the max/min cell diameters (hmax/hmin).

For L one has the Debye lengths (for each species) and the lengths in the geometry. Assuming that the geometry is taken care of by the user in Gmsh, we are only left with the debye length for the spatial criterion:

dx/debye < const

for all species throughout the domain. If it is true for dx_max it is also true throughout the domain. The program cannot correct this by changing the mesh but can issue a warning if it is outside of the threshold.

For T one has both characteristic frequencies and velocities. Characteristic frequencies are both the plasma frequency and the cyclotron frequency. It makes sense to consider them separately, since the pusher is implicit in the B-field, and therefore likely accepts vastly larger time-step for B. We then get two criteria for frequencies (for each species):

dt*wp < const
dt*wc < const

and one for velocities:

dt*v/dx < const

again throughout the domain, and in this case, if it is true for dx_min it is true throughout the domain. As for the characteristic velocity, I use energy conservation. A particle traveling from the outer boundary with speed v0 to the object will end up with a speed v1 given by

v1^2 = v0^2 - 2*q*dphi/m

where dphi is the potential difference the particle passes through. For attracted species v1 is largest, whereas for repelled species v0 is largest. Some particles will end up with a thermal velocity in the same direction as the drift velocity. For these particles,

v0 = vd + vrange*vth

where vrange is an upper cutoff. For pure Dirichlet conditions without magnetic fields, dphi can be taken as the largest positive/negative object potential for negatively/positively charged particles. For floating objects and magnetic fields (when the boundary is not necessarily zero) it is not so trivial, and dphi may perhapse be specified by the user?!

To summarize, this leaves 6 parameters that can be tuned by the user:

max dx/debye
max dt*wp
max dt*wc
max dt*v/dx
vrange
dphi

The program can choose dt to fulfill these criteria and issue a warning if the dx/debye is not satisfactory.

**Comment by sigvaldm** [comment]
All temporal checks (and automatic time-step selection) complete. Will test it on real data tomorrow. Spatial criteria left.
